### PR TITLE
Added config for YOLO life

### PR DESCRIPTION
**Pull request by munsterlander**
Bases on the discussion here: opendatacam#368, I added the calls to the new config options.

### CODE REVIEW

**Comment by munsterlander** [comment]
To clarify here, if a user wants infinite level restarting they can just change the max time to 9999999 or whatever and for users who do not want the box to drop into an infinite loop, they can set it to a reasonable amount or leave the default of 180.

File: config.json [hunk 1]
```diff
@@ -96,5 +96,7 @@
     "hostname": "localhost",
     "signalLossTimeoutSeconds": 60,
     "csvExportOpenStreetMapsUrl": true
-  }
+  },
+  "HTTP_REQUEST_LISTEN_TO_YOLO_RETRY_DELAY_MS":30,
+  "HTTP_REQUEST_LISTEN_TO_YOLO_MAX_RETRIES":180
```
Comment (vsaw) [inline config.json:101]: Do you think we can put this under a YOLO or Darknet group and shorten the name a bit?

Also we should update the documentation to make sure all config options a properly described.

File: server.js [hunk 2]
```diff
@@ -183,6 +183,7 @@ app.prepare()
      */
     express.get('/webcam/stream', (req, res) => {
     // Proxy MJPEG stream from darknet to avoid freezing issues
+      if(!Opendatacam.isListeningToYOLO && mjpgProxy !== null) mjpgProxy = null; //This clears the mjpgProxy object in case YOLO is restarting otherwise a grey screen appears in the UI.
```
Comment (vsaw) [inline server.js:186]: I feel like this only partially fixed the error. 

E.G. what happens if the restart of Yolo finishes before someone refreshes the WebUI. In this case `isListeningToYolo` will again the true and the MJPEG Proxy will not be refreshed. 

In general I wonder if the MJPEG Proxy can detect and fix broken streams on its own. Without having to force restart it.

**Review by vsaw — CHANGES_REQUESTED** [review]

File: config.json (hunk 1 above)
Comment (vsaw) [reply config.json:101]: E.g. we have the Darknet ports further down in the config. Maybe we can combine those in a new group?

Comment (munsterlander) [reply config.json]: I put them under NEURAL_NETWORK_PARAMS as the labels were all related to YOLO.

Comment (munsterlander) [reply config.json]: I mean, we could create a new group if you think that would be better?

File: server.js (hunk 2 above)
Comment (munsterlander) [reply server.js:186]: Well, I am not certain that would happen - but on different hardware maybe?? When I was troubleshooting this, I had debug statements every line so I could trace the exact order (doing breakpoints proved to be ineffective due to the asynchronous nature of some of the operations).  Anyway, consistently it hit that endpoint prior to loading UI, so I do feel it would work no matter what, but I only have the 2 jetsons to test on and I don't have a linux desktop (with Nvidia gpu).

I am not well versed on forever monitor, but that might be a possible solution to monitor the proxy.  Thoughts?

File: config.json (hunk 1 above)
Comment (munsterlander) [reply config.json]: I guess the more I sit here and think about this, if we do the monitor below, then it would probably also need a set of parameters for how long, etc.  Maybe there should be a new group for time parameters or such.  Either way, easy to fix, so let me know the direction you want to go.

File: server.js (hunk 2 above)
Comment (vsaw) [reply server.js:186]: From the point of MJPEG Proxy, all that happens is that the MJPEG Stream gets disconnected. It does not need to know that this is because of Darknet crashing, could just be a Network fault.

Quick tests show that when I crash Darknet manually, the `close` event inside MJPEG Proxy gets raised. However there is no handling of the event. Therefore instead of monitoring Darknet externally I would like to limit the dependencies between ODC and Darknet as much as possible and handle the MJPEG reconnect inside the Proxy.

File: config.json [hunk 3]
```diff
@@ -64,7 +64,9 @@
       "data": "cfg/coco.data",
       "cfg": "cfg/yolov4-tiny.cfg",
       "weights": "yolov4-tiny.weights"
-    }
+    },
+    "retry_delay_ms":30,
+    "max_retries_seconds":180
```
Comment (vsaw) [inline config.json:69]: Like it better this way. I wonder how much effort it will be to group all Darknet related configs here such as `PATH_TO_YOLO_DARKNET`, `CMD_TO_YOLO_DARKNET`, `darknet_json_stream` and `darknet_mjpeg_stream`

**Review by vsaw — CHANGES_REQUESTED** [review]

File: config.json (hunk 1 above)
Comment (vsaw) [reply config.json]: Updating the Markdown files is enough for now. We can worry about docusaurus as we prepare for v4.0.0 release.

And I don't think we should introduce more "timeout" parameters for now. I think people will not care, and indeed be very happy that this is fixed!

File: server.js (hunk 2 above)
Comment (munsterlander) [reply server.js:186]: Ok, that makes sense. Let me look into hooking into the close event.

File: config.json (hunk 1 above)
Comment (munsterlander) [reply config.json]: I'm going to move this out to a generic time group that can be used for any services that may need that.

Comment (munsterlander) [reply config.json]: Documents have been updated and I moved the code to `MONITOR_SETTINGS` so they have their own section in the docs as well as in the config file.

File: config.json (hunk 3 above)
Comment (munsterlander) [reply config.json]: Sorry, I just saw this.  I created a `MONITOR_SETTINGS` group as they didn't really fit anywhere else and they also get their own section in the docs.  As far as grouping darknet config, it shouldn't be too difficult.  I would need to find all the references and update them.  I figure that should be a separate PR for tracking.  I can work on that next week.

File: server.js (hunk 2 above)
Comment (munsterlander) [reply server.js:186]: So regarding hooking into the close event, the issue is the `mjpegProxy` variable.  So here is a thought - again I am completely new to ODC but from what my tests are showing, this seems to be a way forward - we move the declaration of `mjpgProxy` into `Opendatacam.js` so we would reference `Opendatacam.mjpgProxy` wherever its needed.  We import that file into `mjpegProxy.js`, then in the `onClose` event we just need to set `Opendatacam.mjpgProxy` to `null`.  We can get rid of this line `if(!Opendatacam.isListeningToYOLO && mjpgProxy !== null) mjpgProxy = null;` as the actual call to the `/webcam/stream` endpoint would then re-establish the connection.  This should make it more agnostic and aware.  Thoughts on that?

Comment (vsaw) [reply server.js:186]: I don't like this for two reasons

1. Recreating `mjpgProxy` would close all browser clients, and require them to reconnect/refresh the page. This does not seem like a big deal in your case as the UI gets refreshed anyway as Darknet is intentionally restarted, but it will make a difference to situations where Darknet crashes in the back.
2. In the long term I would like to make ODC independent of Darknet to potentially allow plugging in different Engines such as TensorFlow or Deepstream. These engines might not require a MJPEG Proxy. Therefore I would love to keep `Opendatacam.js` independent of Video Streaming and implementation details of the neural network. See for example https://github.com/opendatacam/opendatacam/issues/342 or https://github.com/opendatacam/opendatacam/issues/112 or https://github.com/opendatacam/opendatacam/pull/462

File: config.json [hunk 4]
```diff
@@ -13,6 +13,10 @@
     "remote_hls_gstreamer": "souphttpsrc location=http://YOUR_HLSSTREAM_URL_HERE.m3u8 ! hlsdemux ! decodebin ! videoconvert ! videoscale ! appsink",
     "simulation": "--yolo_json public/static/placeholder/alexeydetections30FPS.json --video_file_or_folder public/static/placeholder/frames --isLive true --jsonFps 20 --mjpgFps 0.2"
   },
+  "MONITOR_SETTINGS": {
+    "retry_delay_ms":30,
+    "max_retries_seconds":180
+  },
```
Comment (vsaw) [inline config.json:19]: "Monitor" seems very generic.

Also these settings as well as e.g. the JSON Ports are very specific to Darknet in my opinion. I therefore wonder if it is best to group them in a Darknet category somehow. e.g.

```JSON
{
    "DARKNET_CONFIG": {
        "retry_delay_ms": 30,
        "max_retries_seconds": 180,
        "json_stream": "http://localhost:8070",
        "mjpeg_stream": "http://localhost:8090"
    }
}
```

I realize that this is part of a bigger refactoring. If you feel like this it to much effort I'm happy to take the PR without the changes in config as I believe not many people will change it anyway.

File: documentation/CONFIG.md [hunk 5]
```diff
@@ -518,6 +519,19 @@ You can tweak some settings of the tracker to optimize OpenDataCam better for yo
...
+#### Monitor settings
+
+```json
+"MONITOR_SETTINGS": {
+    "retry_delay_ms":30,
+    "max_retries_seconds":180
+}
+```
+
+- `retry_delay_ms`: If a monitored service fails, i.e. Darknet YOLO, how long, in milliseconds, should ODC wait before trying to connect again
```
Comment (vsaw) [inline documentation/CONFIG.md:531]: This feels incorrect as e.g. the Mongo Service reconnection is handled differently.

**Review by vsaw — CHANGES_REQUESTED** [review]
The configuration and mjpegProxy still needs fixing. If you wish, feel free to create a new PR just containing your fixes to `isListeningToYolo` in `Opendatacam.js`

**Comment by munsterlander** [comment]
I think its fine for this PR.  We can keep hashing it out because Dev is open to work and the fix is documented here, but if we keep working this out, then its fixed for the better for good.  I spent a good amount of time moving all the darknet path configs and was getting ready to create a PR on that, but based on your above reply, I see where you want to go and will consolidate all of that stuff in prep to decouple darknet from ODC.  That makes sense because on our radar is to migrate to Deepstream as well.

**Comment by vsaw** [comment]
Part of this was merged via https://github.com/opendatacam/opendatacam/pull/513
